src/ray/core_worker/task_execution.h

```diff
@@ -54,26 +58,45 @@ class CoreWorkerTaskExecutionInterface {
   /// just copy their content.
   ///
   /// \param spec[in] Task specification.
-  /// \param args[out] The arguments for passing to task executor.
-  ///
-  Status BuildArgsForExecutor(const TaskSpecification &spec,
-                              std::vector<std::shared_ptr<RayObject>> *args);
+  /// \param args[out] Argument data as RayObjects.
+  /// \param args[out] ObjectIDs corresponding to each by reference argument. The length
+  ///                  of this vector will be the same as args, and by value arguments
+  ///                  will have ObjectID::Nil().
+  ///                  // TODO(edoakes): this is a bit of a hack that's necessary because
+  ///                  we have separate serialization paths for by-value and by-reference
+  ///                  arguments in Python. This should ideally be handled better there.
+  /// \return The arguments for passing to task executor.
+  Status BuildArgsForExecutor(const TaskSpecification &task,
+                              std::vector<std::shared_ptr<RayObject>> *args,
+                              std::vector<ObjectID> *arg_reference_ids);
 
   /// Execute a task.
   ///
   /// \param spec[in] Task specification.
+  /// \param spec[in] Resource IDs of resources assigned to this worker.
   /// \param results[out] Results for task execution.
   /// \return Status.
-  Status ExecuteTask(const TaskSpecification &spec,
+  Status ExecuteTask(const TaskSpecification &task_spec,
+                     const ResourceMappingType &resource_ids,
                      std::vector<std::shared_ptr<RayObject>> *results);
 
+  /// Reference to the parent CoreWorker.
+  /// TODO(edoakes) this is very ugly, but unfortunately necessary so that we
+  /// can clear the ActorHandle state when we start executing a task. Two
+  /// possible solutions are to either move the ActorHandle state into the
+  /// WorkerContext or to remove this interface entirely.
```

Comment:
One option is to expose `CoreWorker::SetCurrentTaskId` and `CoreWorker::SetActorId` to Python and call them from the `execute_task` function. I'm 50/50 though; this seems fine as long as we can get rid of it soon.

Reply:
I considered that too but I think this would add a "gotcha" to the python code (if you don't call those properly/at the right point in the code, things go wrong in somewhat weird ways).